pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/NumericalFilterOptimizer.java

```diff
@@ -0,0 +1,273 @@
+ *
+ * Simple predicate examples:
+ *  1) WHERE "intColumn = 5.0"  gets rewritten to "WHERE intColumn = 5"
+ *  2) WHERE "intColumn != 5.0" gets rewritten to "WHERE intColumn != 5"
+ *  3) WHERE "intColumn = 5.5"  gets rewritten to "WHERE false" because INT values can not match 5.5.
+ *  4) WHERE "intColumn = 3000000000 gets rewritten to "WHERE false" because INT values can not match 3000000000.
+ *  5) WHERE "intColumn != 3000000000 gets rewritten to "WHERE true" becuase INT values always not equal to 3000000000.
+ *
+ * Compound predicate examples:
+ *  6) WHERE "intColumn1 = 5.5 AND intColumn2 = intColumn3"
+ *       rewrite to "WHERE false AND intColumn2 = intColumn3"
+ *       rewrite to "WHERE intColumn2 = intColumn3"
+ *  7) WHERE "intColumn1 != 5.5 OR intColumn2 = 5000000000" (5000000000 is out of bounds for integer column)
+ *       rewrite to "WHERE true OR false"
+ *       rewrite to "WHERE true"
+ *       rewrite to query without any WHERE clause.
+ *
+ * When entire predicate gets rewritten to false (Example 3 above), the query will not return any data. Hence, it is
+ * better for the Broker itself to return an empty response rather than sending the query to servers for further
+ * evaluation.
+ */
+public class NumericalFilterOptimizer implements FilterOptimizer {
+
+  private static final Expression TRUE = RequestUtils.getLiteralExpression(true);
+  private static final Expression FALSE = RequestUtils.getLiteralExpression(false);
+
+  @Override
+  public FilterQueryTree optimize(FilterQueryTree filterQueryTree, @Nullable Schema schema) {
+    // Don't do anything here since this is for PQL queries which we no longer support.
+    return filterQueryTree;
+  }
+
+  @Override
+  public Expression optimize(Expression expression, @Nullable Schema schema) {
+    ExpressionType type = expression.getType();
+    if (type != ExpressionType.FUNCTION) {
+      // Not a function, so we have nothing to rewrite.
+      return expression;
+    }
+
+    Function function = expression.getFunctionCall();
+    List<Expression> operands = function.getOperands();
+    String operator = function.getOperator();
+    if (!(operator.equals(FilterKind.EQUALS.name()) || operator.equals(FilterKind.NOT_EQUALS.name()))) {
+      // This is not an EQUALS or NOT_EQUALS function, but one of its operands may be an EQUALS function so
+      // recursively traverse the expression tree to see if we find an EQUALS function to rewrite.
+      operands.forEach(operand -> optimize(operand, schema));
+
+      // We have rewritten the child operands, so rewrite the parent if needed.
+      return optimizeCurrent(expression);
+    }
+
+    // If we are here, then this expression must have EQUALS operator. Verify that LHS is a numeric column and RHS is
+    // a numeric literal before proceeding further.
+    Expression lhs = operands.get(0), rhs = operands.get(1);
+    if (!(isNumericColumn(lhs, schema) && isNumericLiteral(rhs))) {
```

Comment:
We should always convert integral rhs `DOUBLE` value to `LONG` even if lhs is an expression

Reply:
As we discussed offline earlier, for now we are only rewriting when LHS is column and RHS is a literal as there is quite a bit of complexity involved in determining LHS datatype when LHS is an generic expression. It should be possible, but its probably better to get simpler cases out of the way before stepping into more complex cases.